File: services/admin-service/internal/api/edition.go
package api

import (
	"database/sql"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/vistasecurity/vistaplatform/admin-service/internal/config"
	"github.com/vistasecurity/vistaplatform/admin-service/internal/handlers"
	adminservices "github.com/vistasecurity/vistaplatform/admin-service/internal/services"
	"github.com/vistasecurity/vistaplatform/shared/cache"
)

// This file is the Core/Enterprise seam for admin-service. There are two
// independent Enterprise surfaces behind it: *billing* (monetization) and
// *MSP* (the multi-tenant management plane).
//
// The open-core split is "who is the buyer":
//
//	Core       — the barebones operator console a SINGLE ORGANIZATION needs:
//	             platform users, roles/RBAC and platform-admin auth; platform
//	             settings (SMTP, password/session policy); branding; legal
//	             document authoring; identity providers and staff SSO; storage
//	             config; platform integrations; security events and the
//	             platform's own compliance posture; System Health; and the tier
//	             /billable-item catalog including tier ASSIGNMENT, so
//	             shared/entitlements resolves correctly with no Enterprise code
//	             present.
//	MSP        — the management plane you only need when you operate OTHER
//	             organizations: the whole /admin/tenants/** surface, tenant
//	             lifecycle, per-tenant entitlement overrides, cross-tenant
//	             stats/dashboard/cost aggregates, announcements, maintenance
//	             windows, support tickets, and the cross-tenant legal
//	             acceptance ledger. Mechanically identifiable: it either serves
//	             a /platform aggregate or reads the BYPASSRLS handle to span
//	             tenants.
//	Enterprise — the commercial layer that turns tenants into customers:
//	             Stripe, subscriptions, invoices, coupons, dunning, trials,
//	             contract renewals and billing analytics.
//
// Core declares the seam; the implementation lives under
// services/admin-service/ee/ and is absent from the open-source tree. A Core
// build leaves the hooks nil, so the billing routes are simply never mounted
// and the billing background workers never exist. That is a supported product
// configuration, not a degraded one — nothing in Core reads a Stripe object.
//
// IMPORTANT — what deliberately stays in Core: tier *assignment*.
// adminservices.TierService.AssignTier and the free/manual assignment path in
// handlers/tier_management.go are Core code. A Core deployment can still put a
// tenant on a tier, which is what shared/entitlements resolves against, so
// entitlements work with no billing package present. The only thing Core loses
// is the Stripe Product/Price auto-provisioning that a *stripe-billed* tier
// would trigger on save — see TierPricer below.
//
// See docsv4/internal/operations/OPEN_SOURCE_CARVE_TRACKER.md §5.5 for the
// repo-wide pattern.

// BillingDeps is the plumbing the Enterprise billing surface needs in order to
// mount itself. It is deliberately a small bag of already-constructed
// primitives rather than the *Server: the seam must not widen into "EE can
// reach into anything the server holds".
type BillingDeps struct {
	// DB is the RLS-enforcing pool (crypto_app).
	DB *sql.DB
	// BypassDB is the BYPASSRLS pool (crypto_bypass) used by the deliberately
	// cross-tenant billing reads (platform invoice list, analytics rollups).
	BypassDB *sql.DB
	// Config is the loaded service configuration, including the STRIPE_* keys.
	Config *config.Config

	// AdminGroup is the unauthenticated service root group
	// (/api/v1/admin-service). The tenant-facing billing lanes hang off this
	// and apply their own tenant-cookie auth.
	AdminGroup *gin.RouterGroup
	// Protected is the platform-admin group (/api/v1/admin-service/admin)
	// with JWT auth + StringifyUserID already applied. Permission gates are
	// added per-subgroup by the caller via RequirePlatformPermission.
	Protected *gin.RouterGroup

	// RequirePlatformPermission builds a platform-RBAC gate middleware. Handed
	// over as a function so EE never needs the *rbac.RBACService itself.
	RequirePlatformPermission func(permission string) gin.HandlerFunc

	// CachedHandlers is the Redis-backed handler cache, or nil when Redis is
	// unavailable. EE uses it only to invalidate the tenant cache after a
	// tenant-billing write.
	CachedHandlers *handlers.CachedHandlers
}

// MSPDeps is the plumbing the MSP management plane needs in order to mount
// itself. Same design rule as BillingDeps: a small bag of already-constructed
// primitives, never the *Server.
type MSPDeps struct {
	// DB is the RLS-enforcing pool (crypto_app).
	DB *sql.DB
	// BypassDB is the BYPASSRLS pool (crypto_bypass). Every cross-tenant read
	// in the MSP plane goes through this handle — that is the mechanical tell
	// for what belongs on this side of the seam.
	BypassDB *sql.DB
	// Config is the loaded service configuration. MSP reads
	// EncryptionMasterKey (tenant integration credentials) and the
	// monitoring/resource-tracker peer URLs (dashboard aggregation).
	Config *config.Config

	// Protected is the platform-admin group (/api/v1/admin-service/admin) with
	// JWT auth + StringifyUserID already applied. Permission gates are added
	// per-subgroup by the callee via RequirePlatformPermission.
	Protected *gin.RouterGroup

	// RequirePlatformPermission builds a platform-RBAC gate middleware. Handed
	// over as a function so MSP never needs the *rbac.RBACService itself.
	RequirePlatformPermission func(permission string) gin.HandlerFunc

	// Cache is the Redis client, or nil when Redis is unavailable. MSP uses it
	// to build its own cached cross-tenant reads (tenant list, platform stats);
	// with nil it mounts the uncached handlers instead.
	Cache *cache.Client
	// CachedHandlers is Core's cache janitor, or nil when Redis is
	// unavailable. MSP uses it only for WrapWithCacheInvalidation around its
	// tenant mutations.
	CachedHandlers *handlers.CachedHandlers
}

// MSPRuntime is the handle to MSP background work — currently the dashboard
// aggregator's hub and polling goroutines. Core never has one, so the server's
// start/stop paths are nil-tolerant.
type MSPRuntime interface {
	// Start launches any MSP background workers. Non-blocking.
	Start()
	// Stop signals them to wind down.
	Stop()
}

// BillingRuntime is the handle to the Enterprise billing background workers
// (Stripe webhook processor + contract renewal-notice sweep). Core never has
// one, so the server's start/stop paths are nil-tolerant.
type BillingRuntime interface {
	// Start launches the workers. Non-blocking.
	Start()
	// Stop signals the workers to wind down.
	Stop()
}

// EditionHooks are the extension points the Enterprise build fills in.
//
// The zero value is the Core edition: every hook nil. The Enterprise build
// supplies real implementations from cmd/edition_ee.go, which is guarded by
// `//go:build ee` and is the only file permitted to import
// services/admin-service/ee/. Neither that file nor the ee/ tree exists in the
// open-source repository, so a Core checkout cannot accidentally link
// Enterprise code — there is nothing to link.
//
// Hooks are wired at process start rather than resolved per request: this
// boundary decides which *code* is present, while shared/entitlements decides
// which *tenant* may use it. Both gates apply in an Enterprise build.
type EditionHooks struct {
	// RegisterBilling mounts the whole billing HTTP surface (platform billing
	// admin, tenant /my-billing, onboarding /billing, the provider webhook)
	// and returns a handle to its background workers. Returning nil is legal
	// and means "routes mounted, no workers".
	//
	// Nil in Core, so none of those routes exist.
	RegisterBilling func(BillingDeps) BillingRuntime

	// RegisterMSP mounts the whole MSP management plane (/admin/tenants/**,
	// /admin/stats/**, /admin/dashboard/**, /admin/costs/**,
	// /admin/announcements, /admin/maintenance-windows, /admin/support-tickets,
	// /admin/legal/acceptances, /admin/monitoring/metrics) and returns a handle
	// to its background work. Returning nil is legal and means "routes mounted,
	// no workers".
	//
	// Nil in Core, so a Core console is exactly the single-organization
	// operator surface: it has no tenant directory to show and no cross-tenant
	// aggregate to compute, because that code is not in the binary.
	RegisterMSP func(MSPDeps) MSPRuntime

	// ApplyEditionToken verifies an entitlement token and seeds the grants it
	// carries. Nil in Core, which is why a Core build has no notion of a
	// license token at all — not a disabled check, an absent one.
	//
	// Called once at boot with the platform DB. It must never be fatal: a
	// missing token is the Core edition, and a malformed one is an operator
	// error that should not become an outage.
	ApplyEditionToken func(db *sql.DB)

	// NewTierPricer returns the provisioner that mints a Stripe Product/Price
	// when a stripe-billed tier is saved. Nil in Core: tiers are still fully
	// creatable, editable and assignable, they just carry no Stripe price.
	NewTierPricer func(cfg *config.Config) adminservices.TierPricer
}

// Edition reports the build's edition for startup logging, so an operator can
// tell from the first log line which binary is running.
//
// Both Enterprise surfaces are wired from the same `//go:build ee` file, so in
// practice they are set together; the check is written as "any hook present" so
// it stays honest if that ever changes.
func (h EditionHooks) Edition() string {
	if h.RegisterBilling == nil && h.RegisterMSP == nil {
		return "core"
	}
	return "enterprise"
}

// EditionCapabilities reports which OPTIONAL admin-service surfaces the running
// binary actually mounted. This is the machine-readable half of the edition
// read-out and the half clients should gate on.
//
// It is derived from hook presence, which is the same fact that decides whether
// the routes exist at all — so it cannot drift from reality the way a
// hand-maintained capability list would. `msp` true means /admin/tenants/**,
// /admin/stats/**, /admin/dashboard/**, /admin/costs/**, /admin/announcements,
// /admin/maintenance-windows, /admin/support-tickets and
// /admin/legal/acceptances are mounted; `billing` true means /admin/billing/**
// and /admin/tenants/:id/billing are.
//
// Deliberately NOT a list of every paid capability in the product: this service
// can only speak for its own routes. SIEM export (audit-service/ee/siemexport),
// CMDB sync (inventory-service/ee/cmdbsync) and the rest live in other binaries
// and keep their own response probes — see packages/primitives/src/features/
// edition.ts. Adding them here would be a guess dressed up as a fact.
type EditionCapabilities struct {
	// MSP reports whether the multi-tenant management plane is mounted.
	MSP bool `json:"msp"`
	// Billing reports whether the Enterprise billing surface is mounted.
	Billing bool `json:"billing"`
}

// EditionInfo is the wire shape of GET /admin/platform/edition.
//
// `edition` is deliberately COARSE — "core" or "enterprise", matching
// Edition(). A single ee binary serves both the Enterprise and MSP editions
// (both hooks are wired from the same //go:build ee file), so the binary cannot
// honestly claim to be MSP-licensed rather than Enterprise-licensed; which
// grants a deployment holds is an entitlement-token question, not a build one.
// Clients gate on Capabilities, which IS mechanically exact.
type EditionInfo struct {
	Edition      string              `json:"edition"`
	Capabilities EditionCapabilities `json:"capabilities"`
}

// Info resolves the build's edition read-out.
func (h EditionHooks) Info() EditionInfo {
	return EditionInfo{
		Edition: h.Edition(),
		Capabilities: EditionCapabilities{
			MSP:     h.RegisterMSP != nil,
			Billing: h.RegisterBilling != nil,
		},
	}
}

// PlatformEdition serves the edition read-out at
// GET /api/v1/admin-service/admin/platform/edition.
//
// This handler is CORE code and must stay that way: its whole job is to be
// answerable by a build that has no ee/ tree, so that the admin console can
// stop offering navigation whose backend was never mounted. A Core console that
// shows a Tenants tab is not a licensing problem, it is a broken product — the
// 404 behind that tab is correct behaviour and the tab is the bug.
//
// Authenticated (it hangs off the platform-admin group) but NOT permission
// gated: every operator's navigation depends on it, exactly like
// /admin/auth/me and /admin/user/permissions.
//
// `info` is resolved once at mount time rather than per request because hooks
// are wired at process start and cannot change while the process runs.
func PlatformEdition(info EditionInfo) gin.HandlerFunc {
	return func(c *gin.Context) {
		c.JSON(http.StatusOK, info)
	}
}

File: admin-ui-v2/src/lib/edition.ts
// Platform edition awareness for the operator console.
//
// THE BUG THIS EXISTS TO FIX. admin-service is carved into Core and
// Enterprise/MSP builds. A Core build does not 403 the Enterprise routes — it
// never mounts them, so they 404. Until now the console rendered its whole
// navigation unconditionally, so an operator on a Core deployment clicked
// "Tenants" and got "couldn't load tenants". The 404 was correct; offering the
// tab was the bug.
//
// WHY NOT `useFeature`. The tenant console gates on
// GET /api/v1/auth-service/tenant/features, which resolves a TENANT's
// entitlements. Platform admins live in `platform_users` and have no tenant
// context, so there is no feature map to consult. Edition is also a different
// question from entitlement: entitlement asks "may this customer use it",
// edition asks "is the code even in the binary". This module answers the
// second, from GET /api/v1/admin-service/admin/platform/edition — a Core route,
// answerable by a build with no ee/ tree.
//
// SCOPE. admin-service can only speak for its own routes. Enterprise surfaces
// carved out of OTHER binaries — SIEM export (audit-service/ee/siemexport) —
// keep their own response probes from @vistasecurity/primitives/features. Do
// not extend this map with capabilities this service cannot observe; that would
// be a guess dressed up as a fact.
import { useQuery } from '@tanstack/react-query';
import { clients } from './clients';

/** Optional admin-service surfaces, named as the API reports them. */
export type EditionCapability = 'msp' | 'billing';

export type EditionCapabilities = Record<EditionCapability, boolean>;

/** What the install is LICENSED as (platform_license; an expired licence is core). */
export type LicenseEdition = 'core' | 'enterprise' | 'msp';

/**
 * The licence edition as the console knows it: the reported value, `pending`
 * until the read-out lands, or `unknown` when it could not be read (the
 * request failed, or admin-service reported null).
 */
export type LicenseState = LicenseEdition | 'pending' | 'unknown';

export interface PlatformEdition {
  /** Coarse build edition. Informational — gate on `capabilities`. */
  edition: 'core' | 'enterprise';
  capabilities: EditionCapabilities;
  /**
   * The licence half. The build cannot tell Enterprise from MSP (one ee binary
   * serves both); the licence can. Null when admin-service could not read it.
   * Absent from an admin-service that predates the licence model.
   */
  license_edition?: LicenseEdition | null;
}

/**
 * What we assume before the read-out lands.
 *
 * HIDDEN while loading, on purpose. Either default causes one re-render, so the
 * choice is between the nav growing and the nav shrinking. Growing is the safe
 * direction: an operator can never click an entry that is about to vanish.
 */
export const CAPABILITIES_PENDING: EditionCapabilities = { msp: false, billing: false };

/**
 * What we assume when the read-out cannot be obtained at all.
 *
 * FAIL OPEN, on purpose, and it is the one place in this file worth arguing
 * about. The failure that matters is an admin-service too old to serve the
 * route (it 404s) or briefly unreachable. Failing closed there would blank half
 * of a paying MSP deployment's console over a transient error — a new, worse
 * outage. Failing open degrades to exactly the pre-fix behaviour: the tab is
 * offered and, if this really is Core, the section's own load fails. Bad, but
 * it is the status quo rather than a regression.
 */
export const CAPABILITIES_UNKNOWN: EditionCapabilities = { msp: true, billing: true };

export const platformEditionKey = ['platform', 'edition'] as const;

/**
 * Query options for the edition read-out. Exported separately from the hook so
 * tests can drive it through a real QueryClient.
 *
 * The capability half cannot change while the console is open (hooks are wired
 * at process start); the licence half can — a licence is installed, renewed or
 * lapses while the process runs — so the answer goes stale after five minutes
 * and is re-read on the next navigation. Still one request per mount otherwise.
 */
export function platformEditionQuery() {
  return {
    queryKey: platformEditionKey,
    queryFn: async (): Promise<PlatformEdition> => {
      const { data, error } = await clients.admin.GET('/admin/platform/edition', {});
      if (error || !data) throw new Error('Failed to resolve platform edition');
      return data as PlatformEdition;
    },
    staleTime: 5 * 60 * 1000,
    gcTime: Infinity,
    // An absent route is a settled fact; one retry covers a transient blip
    // without multiplying 404s in the console.
    retry: 1,
  };
}

export interface PlatformEditionState {
  /** Resolved capabilities, already collapsed through the pending/unknown rules. */
  capabilities: EditionCapabilities;
  /** True once the read-out has settled either way. */
  resolved: boolean;
  /** The reported edition, or null until it settles. */
  edition: PlatformEdition['edition'] | null;
  /** Convenience predicate for nav filters and route guards. */
  has: (capability: EditionCapability) => boolean;
  /** The licence edition, or pending/unknown. See LicenseState. */
  license: LicenseState;
  /**
   * True only when the licence is known to be MSP: plans, trials and billing
   * are MSP-only (edition-licensing spec §1). Unknown fails OPEN here, like the
   * capability map, so a transient read failure never blanks a paying MSP's
   * billing surfaces; pending hides them until the answer lands.
   */
  isMsp: boolean;
}

/**
 * Resolve which optional admin-service surfaces this deployment actually has.
 *
 * Safe to call from many components — react-query dedupes to a single request
 * and the result never goes stale.
 */
export function usePlatformEdition(): PlatformEditionState {
  const { data, isError } = useQuery(platformEditionQuery());
  return resolveEditionState(data, isError);
}

/**
 * The pure core of `usePlatformEdition`, so the pending / failed / resolved
 * rules can be tested without mounting React.
 */
export function resolveEditionState(
  data: PlatformEdition | undefined,
  isError: boolean,
): PlatformEditionState {
  const capabilities = data?.capabilities ?? (isError ? CAPABILITIES_UNKNOWN : CAPABILITIES_PENDING);
  const license: LicenseState = data ? (data.license_edition ?? 'unknown') : isError ? 'unknown' : 'pending';
  return {
    capabilities,
    resolved: Boolean(data) || isError,
    edition: data?.edition ?? null,
    has: (capability) => capabilities[capability] !== false,
    license,
    isMsp: license === 'msp' || license === 'unknown',
  };
}

/** Human label for the edition badge in the shell. */
export function editionLabel(edition: PlatformEdition['edition'] | null): string | null {
  if (edition === 'core') return 'Core';
  if (edition === 'enterprise') return 'Enterprise';
  return null;
}
